src/coprocessor/cache.rs

```diff
@@ -0,0 +1,490 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::option::Option;
+use std::boxed::Box;
+use std::sync::{Arc, Mutex};
+use linked_hash_map::LinkedHashMap;
+use super::metrics::*;
+
+type DistSQLCacheKey = String;
+
+const DISTSQL_CACHE_ENTRY_ADDITION_SIZE: usize = 40;
+
+pub struct DistSQLCacheEntry {
+    region_id: u64,
+    version: u64,
+    result: Vec<u8>,
+    size: usize,
+}
+
+impl DistSQLCacheEntry {
+    pub fn new(region_id: u64, version: u64, key_size: usize, r: Vec<u8>) -> DistSQLCacheEntry {
+        let size = r.len() + DISTSQL_CACHE_ENTRY_ADDITION_SIZE + 2 * key_size;
+        DistSQLCacheEntry {
+            region_id: region_id,
+            version: version,
+            result: r,
+            size: size,
+        }
+    }
+}
+
+pub struct RegionDistSQLCacheEntry {
+    version: u64,
+    enable: bool,
+    cached_items: HashMap<DistSQLCacheKey, u8>,
+}
+
+
+#[derive(Default)]
+pub struct DistSQLCache {
+    regions: HashMap<u64, RegionDistSQLCacheEntry>,
+    max_size: usize,
+    map: LinkedHashMap<DistSQLCacheKey, Box<DistSQLCacheEntry>>,
```

Comment:
Seems we do not need a box to wrap `DistSQLCacheEntry `   since the  `LinkedHashMap` already wrapped the value with box https://contain-rs.github.io/linked-hash-map/src/linked_hash_map/lib.rs.html#314